testsuite/model/src/test/java/org/keycloak/testsuite/model/transaction/util/TransactionController.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.testsuite.model.transaction.util;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.KeycloakTransactionManager;
+
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+public class TransactionController {
```

Comment:
I had a look at this new class and don't understand why it creates an Executor and runs everything in its own thread - the Keycloak session is not bound to a thread. The tests seems to complete with this simpler version as well:

```
public class TransactionController {
    private final KeycloakSession session;

    public TransactionController(KeycloakSessionFactory sessionFactory) {
        session = sessionFactory.create();
    }

    public void begin() {
        getTransactionManager().begin();
    }

    public void commit() {
        getTransactionManager().commit();
    }

    public void rollback() {
        getTransactionManager().rollback();
    }

    public <R> R runStep(Function<KeycloakSession, R> task) {
        return task.apply(session);
    }

    private KeycloakTransactionManager getTransactionManager() {
        return session.getTransactionManager();
    }

}
```

Reply:
Using different threads for each transaction should resemble a real scenario where each request is processed in a different thread.

Also, not sure about JPA but I believe this is necessary for HotRod as it refuses to start more transactions from the same thread.

Reply:
If it proofs that it is necessary for HotRod, I'd like to see a comment stating that. 

When keeping the Executor, a small simplification should be possible to change the `static ThreadLocal<KeycloakSession>` to an instance variable `KeycloakSession`

Reply:
I will test your suggestion with ISPN and update the class accordingly.

Reply:
Another note when we keep the Executor: It needs to be shut down at the end of the test execution, otherwise it will leak threads. If you assign it to an "ExecutorService" instead of an "Executor", you can shut it down.

Reply:
So I tested it with the HotRod transaction and it works as you suggested. I was probably doing something differently when ISPN refused to create more transactions for one thread.

On the other hand, I quite like the idea of the steps running in different threads as I think it is closer to a real scenario. What is your opinion on this?

Reply:
I'd like to start as simple as possible ([YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)), and add the complexity when there is the need for it. In this case, I'd vote for removing the thread pool.

Reply:
Done